editor/store/selectors.js

```diff
@@ -1202,152 +1202,268 @@ export function getNotices( state ) {
- * @param {Object}           blockType         Block type, likely from getBlockType().
+ * @param {Object} state      Global application state.
+ * @param {string} blockName  The name of the given block type, e.g. 'core/paragraph'.
+ * @param {?string} parentUID The parent that the given block is to be nested within, or null.
  *
- * @return {Editor.InserterItem} Item that appears in inserter.
+ * @return {boolean} Whether or not the given block type is allowed to be inserted.
  */
-function buildInserterItemFromBlockType( state, allowedBlockTypes, blockType ) {
-	if ( ! allowedBlockTypes || ! blockType ) {
-		return null;
-	}
+export const canInsertBlockType = createSelector(
+	( state, blockName, parentUID = null ) => {
+		const checkAllowList = ( list, item, defaultResult = null ) => {
+			if ( isBoolean( list ) ) {
+				return list;
+			}
+			if ( isArray( list ) ) {
+				return includes( list, item );
+			}
+			return defaultResult;
+		};
 
-	const blockTypeIsDisabled = Array.isArray( allowedBlockTypes ) && ! includes( allowedBlockTypes, blockType.name );
-	if ( blockTypeIsDisabled ) {
-		return null;
-	}
+		const blockType = getBlockType( blockName );
+		if ( ! blockType ) {
+			return false;
+		}
 
-	if ( blockType.isPrivate ) {
-		return null;
-	}
+		const editorAllowedBlocks = getEditorSettings( state ).allowedBlockTypes;
+		const isBlockAllowedInEditor = checkAllowList( editorAllowedBlocks, blockName, true );
+		if ( ! isBlockAllowedInEditor ) {
+			return false;
+		}
 
-	return {
-		id: blockType.name,
-		name: blockType.name,
-		initialAttributes: {},
-		title: blockType.title,
-		icon: blockType.icon,
-		category: blockType.category,
-		keywords: blockType.keywords,
-		isDisabled: !! blockType.useOnce && getBlocks( state ).some( ( block ) => block.name === blockType.name ),
-	};
-}
+		const parentBlockListSettings = getBlockListSettings( state, parentUID );
+		const parentAllowedBlocks = get( parentBlockListSettings, [ 'supportedBlocks' ] );
+		const hasParentAllowedBlock = checkAllowList( parentAllowedBlocks, blockName );
+
+		const blockAllowedParentBlocks = blockType.allowedParentBlocks;
+		const parentName = getBlockName( state, parentUID );
+		const hasBlockAllowedParent = checkAllowList( blockAllowedParentBlocks, parentName );
+
+		let isBlockAllowedInParent;
+		if ( hasParentAllowedBlock !== null && hasBlockAllowedParent !== null ) {
+			isBlockAllowedInParent = hasParentAllowedBlock || hasBlockAllowedParent;
+		} else if ( hasParentAllowedBlock !== null ) {
+			isBlockAllowedInParent = hasParentAllowedBlock;
+		} else if ( hasBlockAllowedParent !== null ) {
+			isBlockAllowedInParent = hasBlockAllowedParent;
+		} else {
+			isBlockAllowedInParent = true;
+		}
+
+		if ( ! isBlockAllowedInParent ) {
+			return false;
+		}
+
+		return true;
+	},
+	( state, blockName, parentUID ) => [
+		state.blockListSettings[ parentUID ],
+		state.editor.present.blocksByUID[ parentUID ],
+		state.settings,
```

Comment:
I think we may be more granular when passing our cache variables and use the selectors to compute the references e.g:[ getEditorSettings( state ).allowedBlockTypes, getBlockListSettings( state, parentUID )... ] this way if the state changes as long as selectors we depend on are updated we are safe :)

Reply:
Good catch. I've updated the `settings` dependency to be more specific in a60dc232e.

I'm not sure it's a good idea to call a selector from the second argument of `createSelector`. If we invalidated the cache based using `getBlockListSettings( state, parentUID )` and then someone went ahead and changed `getBlockListSettings` to use `createSelector` then it would mean that our cache would never invalidate.

It looks like @aduth implemented a really good solution to this problem in the form of [`getDependants`](https://github.com/aduth/rememo#api) but we are not using version 3.0.0 of rememo yet.

I'll look into updating our version of rememo in a seperate PR so that we can clean this up.